client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -203,8 +217,41 @@ private Checkpoint completeCheckpoint(String checkpointName) {
 
     @Override
     public void resetReaderGroup(ReaderGroupConfig config) {
-        Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, config);
-        synchronizer.updateStateUnconditionally(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(config)));
+        while (true) {
+            synchronizer.fetchUpdates();
+            val currentConfig = synchronizer.getState().getConfig();
+            // We only move into the block if the state transition has happened successfully.
+            if (stateTransition(currentConfig, new UpdatingConfig(true))) {
+                // Use the latest generation and reader group Id.
+                ReaderGroupConfig newConfig = config.toBuilder()
+                        .readerGroupId(currentConfig.getReaderGroupId())
+                        .generation(currentConfig.getGeneration()).build();
+                boolean success = Futures.getThrowingException(controller.updateReaderGroup(scope, groupName, newConfig));
+                if (success) {
+                    Map<SegmentWithRange, Long> segments = getSegmentsForStreams(controller, newConfig);
+                    synchronizer.updateState((s, updates) -> {
+                        updates.add(new ReaderGroupStateInit(config, segments, getEndSegmentsForStreams(newConfig), false));
```

Comment:
This is storing `config` which was passed as a parameter. It does not appear to have a generation set on it.

What I see happening now is that resetReaderGroup is called with a config provided by the user. Presumably they don't set the generation. (Should that method be private?) So the default value of -1 is passed. The call ends up here and it passes the config into the state init. But in `updateConfigIfNeeded` it pulls from the controller which updates the generation id every time. So it will always return a new config. Thus we end up reinitializing the group state twice.

Reply:
`controller.updateReaderGroup` now returns the updated generation( See PR: https://github.com/pravega/pravega/pull/5462), which should be written to StateSynchronizer, but I don't see that happening here.

Reply:
I have made it such that the updated generation is returned. That generation is what is written into the `StateSynchronizer`.